In [3]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Importing products
df_prods = pd.read_csv(r"C:\Users\abiya\01-2023 Instacart Basket Analysis\Data\Original Data\products.csv", index_col = False)

In [4]:
#Importing orders
df_ords = pd.read_csv(r"C:\Users\abiya\01-2023 Instacart Basket Analysis\Data\Prepared Data\orders_wrangled.csv", index_col = False)

In [4]:
#Checking consistensy of the products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [27]:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


#There is mixed data type in the product_name column and addressed by changing it to string since it is name.


In [28]:
#Chaning data type to string
df_prods['product_name'] = df_prods['product_name'].astype('str')

In [29]:
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

#No missing values

In [31]:
df_dups = df_prods[df_prods.duplicated()]

#No duplicated value found.

In [7]:
#Checking consistensy of the wrangled orders dataframe 
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#Clsrification
##No weird recording is observed. There are only seven days and 24 hours a week and a day respectively, the dataframe has them accordingly. There is no also negative representation in the days since prior order. So it is good so far. Having said this, searching for missing and duplicate data will be what we have to worry about.



In [10]:
#Checking for mixed type columns in orders dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#Since the dataframe was ready for analysis no mixed type column is observed.

In [36]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#There are 206209 missing observations in the days since prior order column.

In [5]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [6]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

#Solution:- Since all of the observations containing missing values in the 'days_since_prior_order' have an 'order_number' valued as 1 indicating the customer is ordering for the first time and it won't be a value showing how long it has been since a prior order, not making any change is preferable. I have chosen not to make any changes to the missing values in the 'days_since_prior_order' column. Rather, a value of 1 in the 'order_number' column can act as a flag for a new customer.

In [10]:
df_ords['days_since_prior_order_flag'] = df_ords['days_since_prior_order']

In [13]:
df_ords['days_since_prior_order_flag'] == df_ords['days_since_prior_order']

0          False
1           True
2           True
3           True
4           True
           ...  
3421078     True
3421079     True
3421080     True
3421081     True
3421082     True
Length: 3421083, dtype: bool

In [16]:
df_ords['order_number'] == 1

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Name: order_number, Length: 3421083, dtype: bool

In [20]:
#Creating a flag
df_ords.loc[df_ords['order_number'] == 1, 'days_since_prior_order_flag'] = 'First order'

In [21]:
df_ords['days_since_prior_order_flag']

0          First order
1                 15.0
2                 21.0
3                 29.0
4                 28.0
              ...     
3421078           29.0
3421079           30.0
3421080           18.0
3421081            7.0
3421082           30.0
Name: days_since_prior_order_flag, Length: 3421083, dtype: object

In [23]:
#Checking for duplicated observations
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [26]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#No duplicated observation found.

In [32]:
#Exporting files to prepared folder
path = r"C:\Users\abiya\01-2023 Instacart Basket Analysis"

In [39]:
df_prods.to_csv(os.path.join(path, 'Data','Prepared Data', 'prods_cleaned.csv'))

In [40]:
df_ords_clean.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_cleaned.csv'))

#Correction made based on the provided direction. 